File: apps/mobile/src/models/Allowance.ts
import { Model } from '@nozbe/watermelondb';
import { field, date, readonly, relation } from '@nozbe/watermelondb/decorators';
import type { Associations } from '@nozbe/watermelondb/Model';
import type Family from './Family';
import type FamilyMember from './FamilyMember';
import type Wallet from './Wallet';

export type AllowanceFrequency = 'daily' | 'weekly' | 'biweekly' | 'monthly' | 'custom';
export type AllowanceStatus = 'active' | 'paused' | 'archived';

export default class Allowance extends Model {
  static table = 'allowances';

  static associations: Associations = {
    families: { type: 'belongs_to', key: 'family_id' },
    wallets: { type: 'belongs_to', key: 'wallet_id' },
  };

  @field('family_id') familyId!: string;
  @field('giver_member_id') giverMemberId!: string;
  @field('receiver_member_id') receiverMemberId!: string;
  @field('wallet_id') walletId!: string;
  @field('amount_cents') amountCents!: number;
  @field('frequency') frequency!: AllowanceFrequency;
  @field('interval_count') intervalCount!: number;
  @date('next_run_at') nextRunAt?: Date;
  @date('last_run_at') lastRunAt?: Date;
  @field('status') status!: AllowanceStatus;
  @field('notes') notes?: string;

  @readonly @date('created_at') createdAt!: Date;
  @readonly @date('updated_at') updatedAt!: Date;

  @relation('families', 'family_id') family!: Family;
  @relation('wallets', 'wallet_id') wallet!: Wallet;
}

File: src-tauri/src/models/allowance.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum Frequency {
    Daily,
    Weekly,
    Biweekly,
    Monthly,
    Custom,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum AllowanceStatus {
    Active,
    Paused,
    Archived,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct Allowance {
    pub id: i64,
    pub uuid: String,
    pub family_id: i64,
    pub giver_member_id: i64,
    pub receiver_member_id: i64,
    pub wallet_id: i64,
    pub amount_cents: i64,
    pub frequency: String,
    pub interval_count: i64,
    pub next_run_at: Option<String>,
    pub last_run_at: Option<String>,
    pub status: String,
    pub notes: Option<String>,
    pub sync_version: i64,
    pub last_synced_at: Option<String>,
    pub is_deleted: i64,
    pub created_at: String,
    pub updated_at: String,
}
